tools/schema_loader.cc

```diff
@@ -649,6 +649,89 @@ schema_ptr do_load_schema_from_schema_tables(const db::config& dbcfg, std::files
     }
 }
 
+schema_ptr do_load_schema_from_sstable(const db::config& dbcfg, std::filesystem::path sstable_path, sstring keyspace, sstring table) {
+    if (keyspace.empty()) {
+        keyspace = "my_keyspace";
+    }
+    if (table.empty()) {
+        table = "my_table";
+    }
+
+    db::nop_large_data_handler large_data_handler;
+    gms::feature_service feature_service(gms::feature_config_from_db_config(dbcfg));
+    cache_tracker tracker;
+    sstables::directory_semaphore dir_sem(1);
+    sstables::sstables_manager sst_man("tools::load_schema_from_sstable", large_data_handler, dbcfg, feature_service, tracker,
+        memory::stats().total_memory(), dir_sem,
+        [host_id = locator::host_id::create_random_id()] { return host_id; });
+    auto close_sst_man = deferred_close(sst_man);
+
+    schema_ptr bootstrap_schema = schema_builder(keyspace, table).with_column("pk", int32_type, column_kind::partition_key).build();
+
+    const auto ed = sstables::parse_path(sstable_path, keyspace, table);
+    const auto dir_path = sstable_path.parent_path();
+    data_dictionary::storage_options local;
+    auto bootstrap_sst = sst_man.make_sstable(bootstrap_schema, dir_path.c_str(), local, ed.generation, sstables::sstable_state::normal, ed.version, ed.format);
+
+    bootstrap_sst->load_metadata().get();
+
+    const auto& serialization_header = bootstrap_sst->get_serialization_header();
+    const auto& compression = bootstrap_sst->get_compression();
+
+    auto builder = schema_builder(keyspace, table);
+
+    const auto to_string_view = [] (bytes_view b) {
+        return std::string_view(reinterpret_cast<const char*>(b.data()), b.size());
+    };
+    const auto parse_type = [&to_string_view] (bytes_view type_name) {
+        return db::marshal::type_parser::parse(to_string_view(type_name));
+    };
+
+    // partition key
+    {
+        const auto pk_type_name = serialization_header.pk_type_name.value;
+
+        const bytes composite_prefix = "org.apache.cassandra.db.marshal.CompositeType(";
+
+        if (pk_type_name.starts_with(composite_prefix)) {
+            const auto composite_content = pk_type_name.substr(composite_prefix.size(), pk_type_name.size() - composite_prefix.size() - 1);
+
+            db::marshal::type_parser parser(to_string_view(composite_content));
+            unsigned i = 0;
+
+            while (!parser.is_eos()) {
+                builder.with_column(to_bytes(format("pk{}", i++)), parser.parse(), column_kind::partition_key);
```

Comment:
This will fail if there is a regular column named "pk0" (which of course there will be the first time we try this). Add "$" prefix to mark it as synthetic.

Reply:
Good point, will do.

Reply:
Fixed in v2.